articles/actions.md

```diff
@@ -0,0 +1,338 @@
+
+### Goal Identifiers
+
+In ROS 1, Action clients are responsible for creating a goal ID when submitting a goal.
+In ROS 2 the action server will be responsible for generating the goal ID and notifying the client.
+
+The server is better equiped to generate a unique goal id than the client because there may be multiple clients who could independently generate the same goal id.
+Another reason is to avoid a race condition between goal creation and cancellation that exists in ROS 1.
+In ROS 1 if a client submits a goal and immediatly tries to cancel it then the cancelation may or may not happen.
+If the cancelation is processed before the goal submission then `actionlib` will ignore the cancellation request without notifying the user's code.
+
+### Namespacing of Generated Messages and Services
+
+Multiple message and service definitions are generated from a single action definition.
+In ROS 1, the generated messages were prefixed with the name of the action to avoid conflicts with other messages and services.
+In ROS 2, the generated service and message definitions will exist in a different namespace to be impossible to conflict with non-action message and service definitions.
+For example, in Python the code from the generated definitions should be in the module `action` instead of `srv` and `msg`.
+In C++, the generated code should be in the namespace and folder `action` instead of `srv` and `msg`.
+
+### Visibility of Action Services and Topics
+
+In ROS 1 `rostopic list` would show all action topics in its output.
+In ROS 2 `ros2 topic list` and `ros2 service list` will not show topics and services used by actions by default.
+The can still be shown by passing an option to the commands to show hidden services and topics.
+
+## Action Interface Definition
+
+Actions are specified using a form of the ROS Message IDL.
+The specification contains three sections, each of which is a message specification:
+
+1. Goal
+
+   This describes what the action should achieve and how it should do it.
+   It is sent to the action server when it is requested to execute an action.
+
+1. Result
+
+   This describes the outcome of an action.
+   It is sent from the server to the client when the action execution ends, whether successfully or not.
+
+1. Feedback
+
+   This describes the progress towards completing an action.
+   It is sent to the client of the action from the action server between commencing action execution and prior to the action completing.
+   This data is used by the client to understand the progress of executing the action.
+
+Any of these sections may be empty.
+Between each of the three sections is a line containing three hyphens, `---`.
+Action specifications are stored in a file ending in `.action`.
+There is one action specification per `.action` file.
+
+### Example
+
+```
+# Define a goal of washing all dishes
+uint32 dishwasher_id  # Specify which dishwasher we want to use
```

Comment:
i don't like this example if its going to be *the* example, because the solo ``dishwasher_id`` may be confused with the action ID. how about add another goal input e.g. `bool heavy_duty # spend extra time cleaning`